### PR TITLE
Session storage and retrieval store APIs(redis)

### LINKED ISSUE #336: Implement session storage and retrieval store APIs (Redis)
This is the Redis in-memory store, which will store persistent details about on-going sessions, such as parameters that should be collected by the parameter collection module and whether an on-going parameter collection session is going on. This will replace the currently used Rasa slot management service.

### PR DESCRIPTION
**Pull request by nuwangeek**
This pull request introduces a Redis-backed session store for managing multi-turn API tool workflows, including all necessary infrastructure, application logic, configuration, and documentation. The main changes add a robust, async CRUD interface for storing session state in Redis, integrate the store into the FastAPI app lifecycle, and provide clear usage documentation. Redis connectivity is now surfaced in the service health check, and all configuration is parameterized for flexible deployment.

**Session Store Implementation and Integration**
- Added `APIToolSessionStore` and `APIToolSession` Pydantic model to enable async CRUD operations for agentic session state, with sliding TTL and partial updates. (`src/utils/api_tool_session_store.py`, `src/models/session_models.py`) [[1]](diffhunk://#diff-a002c020fc30215a470106304e9c1659a22813c017b1fe8a877cfa308c917b13R1-R126) [[2]](diffhunk://#diff-ee6408ce1fb7cf5d762fe7461ea33c55fa4fc98a45edc4176958e0be961eea31R1-R43)
- Integrated session store initialization and shutdown into the FastAPI app’s lifespan events, making the store available at `app.state.session_store` and handling Redis outages gracefully. (`src/llm_orchestration_service_api.py`) [[1]](diffhunk://#diff-04c83e29b0e1f92c363ce770fee0480e32ff506036da7faa6632ddef2065b0c3R15-R20) [[2]](diffhunk://#diff-04c83e29b0e1f92c363ce770fee0480e32ff506036da7faa6632ddef2065b0c3R90-R98) [[3]](diffhunk://#diff-04c83e29b0e1f92c363ce770fee0480e32ff506036da7faa6632ddef2065b0c3R110-R111)
- Health check endpoint now reports Redis session store connectivity status. (`src/llm_orchestration_service_api.py`)

**Infrastructure and Configuration**
- Added Redis connection management utilities, including async client initialization, teardown, and health checking, with environment-variable-based configuration for host, port, password, and DB selection. (`src/utils/redis_client.py`, `env.example`) [[1]](diffhunk://#diff-d9bfdb44b8bc09de5fbc8c20a821297034f2a41956f44ed167f8a2c9d0e79791R1-R76) [[2]](diffhunk://#diff-0521255bbda96fd681ef7be9c0d23f04cb3838b3a905ed7f25b041ec34cf547cR65)
- Updated Docker Compose files to inject Redis connection settings as environment variables, ensure the LLM orchestration service depends on Redis, and expose configuration for the session DB. (`docker-compose.yml`, `docker-compose-ec2.yml`) [[1]](diffhunk://#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3R580-R583) [[2]](diffhunk://#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3R593) [[3]](diffhunk://#diff-02a58319624c828920f29b420fb9c70bf5b85a46a9200a13981d1d98b86ba08dR633-R636) [[4]](diffhunk://#diff-02a58319624c828920f29b420fb9c70bf5b85a46a9200a13981d1d98b86ba08dR645)

**Documentation and Dependencies**
- Added a comprehensive usage guide for the Redis session store, including example CRUD patterns and schema reference. (`docs/REDIS_SESSION_STORE.md`)
- Declared `redis[hiredis]` as an application dependency and updated lint configuration to allow dynamic field updates in the session store. (`pyproject.toml`) [[1]](diffhunk://#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R40) [[2]](diffhunk://#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R114-R115)

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -577,6 +577,10 @@ services:
       - ENVIRONMENT=production
       - VAULT_ADDR=http://vault-agent-llm:8201
       # VAULT_TOKEN not set - vault-agent-llm proxy handles authentication
+      - REDIS_HOST=${REDIS_HOST:-redis}
+      - REDIS_PORT=${REDIS_PORT:-6379}
+      - REDIS_AUTH=${REDIS_AUTH:-myredissecret}
```
Comment (Thirunayan22) [inline docker-compose.yml]: **[CRITICAL] Hardcoded default Redis password**

`myredissecret` as a default means any deployment that forgets to set `REDIS_AUTH` runs with a known password. This violates **ADR SEC-001** — secrets must never be hardcoded and should be injected via Vault/K8s Secrets at runtime.

**Why it matters:** This password is publicly visible in the repo. An attacker with network access to Redis can read/write all session data.

**Suggested fix:** Remove the default so the service fails fast if unset:
```yaml
- REDIS_AUTH=${REDIS_AUTH:?REDIS_AUTH must be set}
```
Or add a startup guard that rejects known-weak defaults in production.

> Same issue exists in `docker-compose-ec2.yml:636`.

File: src/utils/api_tool_session_store.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        """
+        session = await self.get(chat_id)
+        if session is None:
+            logger.warning(
+                "[SessionStore] update({}) - session not found, skipping", chat_id
+            )
+            return None
+
+        updated = session.model_copy(update=fields)
+        await self.save(updated)
```
Comment (Thirunayan22) [inline src/utils/api_tool_session_store.py]: **[CRITICAL] Non-atomic read-modify-write race condition**

`update()` does `GET` → modify in Python → `SET`. Between the `GET` and `SET`, a concurrent request for the same `chat_id` can overwrite this session, losing the first request's changes.

**Why it matters:** Concurrent retries, SSE reconnects, or webhook callbacks on the same chat could silently drop collected parameters or reset turn counts.

**Suggested fix:** Use a Lua script for atomic read-modify-write:
```python
_LUA_MERGE = """
local raw = redis.call('GET', KEYS[1])
if not raw then return nil end
local session = cjson.decode(raw)
local updates = cjson.decode(ARGV[2])
for k, v in pairs(updates) do session[k] = v end
local merged = cjson.encode(session)
redis.call('SET', KEYS[1], merged, 'EX', tonumber(ARGV[1]))
return merged
"""
```
Or use `WATCH`/`MULTI`/`EXEC` for optimistic locking.

File: src/utils/redis_client.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+def _build_redis_url() -> str:
+    """Build Redis URL from environment variables."""
+    host = os.getenv("REDIS_HOST", "redis")
+    port = os.getenv("REDIS_PORT", "6379")
+    password = os.getenv("REDIS_AUTH", "")
+    db = os.getenv("REDIS_SESSION_DB", "1")
+
+    if password:
+        return f"redis://:{password}@{host}:{port}/{db}"
+    return f"redis://{host}:{port}/{db}"
```
Comment (Thirunayan22) [inline src/utils/redis_client.py]: **[HIGH] No TLS support — credentials sent in plaintext**

This builds `redis://` URLs only. The project already defines `REDIS_TLS_ENABLED`, `REDIS_TLS_CA`, `REDIS_TLS_CERT`, `REDIS_TLS_KEY` in `env.example`, but this client ignores them entirely.

**Why it matters:** In non-localhost deployments (especially the EC2 compose), the Redis password travels over the network unencrypted.

**Suggested fix:**
```python
def _build_redis_url() -> str:
    tls = os.getenv("REDIS_TLS_ENABLED", "false").lower() == "true"
    scheme = "rediss" if tls else "redis"
    ...
    return f"{scheme}://:{password}@{host}:{port}/{db}"
```
And pass `ssl_ca_certs`, `ssl_certfile`, `ssl_keyfile` to `from_url()` when TLS is enabled.

File: src/llm_orchestration_service_api.py [hunk 4]
```diff
@@ -81,6 +87,15 @@ async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
...
         raise
 
+    # Initialize Redis session store (non-fatal: service continues without it)
+    try:
+        await init_redis_client()
+        app.state.session_store = APIToolSessionStore()
+        logger.info("Redis session store initialized successfully")
+    except Exception as e:
+        logger.warning(f"Redis session store unavailable, continuing without it: {e}")
+        app.state.session_store = None
```
Comment (Thirunayan22) [inline src/llm_orchestration_service_api.py:97]: **[HIGH] Silent `None` degrades multi-turn workflows with no user signal**

When Redis is down, `session_store` is `None` and every consumer must null-check. The multi-turn API tool workflow silently breaks — users experience conversations that "forget" context with no error indication.

**Why it matters:** This is a silent data-loss path. Consumers who forget the `None` check get `AttributeError` at runtime.

**Suggested fix:** When a request hits a code path that *requires* the session store and it's `None`, return HTTP 503 with a clear message rather than silently degrading. The health endpoint already reports the status, so monitoring can catch it.

File: src/utils/redis_client.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
+"""Redis async connection manager for session store."""
+
+import os
+from typing import Optional
+
+import redis.asyncio as aioredis
+from loguru import logger
+
+
+_redis_client: Optional[aioredis.Redis] = None  # type: ignore[type-arg]
```
Comment (Thirunayan22) [inline src/utils/redis_client.py:10]: **[HIGH] Module-level global mutable state**

`_redis_client` as a module-global mutated via `init`/`close` is not concurrency-safe and makes testing harder. If `close_redis_client()` is called during shutdown while a request is in-flight, `get_redis_client()` returns `None` mid-operation.

**Why it matters:** Potential `NoneType` errors during graceful shutdown under load.

**Suggested fix:** Store the client on `app.state.redis_client` and pass it through FastAPI's dependency injection. This aligns with how `orchestration_service` is already managed in this file.

File: src/utils/api_tool_session_store.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            await client.set(
+                _key(session.chat_id),
+                session.model_dump_json(),
+                ex=_SESSION_TTL_SECONDS,
+            )
+            logger.debug("[SessionStore] Session saved for chat_id={}", session.chat_id)
+        except Exception as exc:
+            logger.error("[SessionStore] save({}) failed: {}", session.chat_id, exc)
+
+    async def update(self, chat_id: str, **fields: Any) -> Optional[APIToolSession]:
```
Comment (Thirunayan22) [inline src/utils/api_tool_session_store.py:76]: **[MEDIUM] `**fields: Any` accepts arbitrary kwargs without validation**

Typos like `sate="ready"` instead of `state="ready"` are silently ignored by `model_copy(update=...)`. The caller gets no feedback that their update did nothing.

**Suggested fix:** Add a guard:
```python
valid = set(APIToolSession.model_fields)
unknown = set(fields) - valid
if unknown:
    raise ValueError(f"Unknown session fields: {unknown}")
```

File: env.example [hunk 7]
```diff
@@ -62,6 +62,7 @@ REDIS_TLS_ENABLED=false
 REDIS_TLS_CA=/certs/ca.crt
 REDIS_TLS_CERT=/certs/redis.crt
 REDIS_TLS_KEY=/certs/redis.key
+REDIS_SESSION_DB=1
```
Comment (Thirunayan22) [inline env.example:65]: **[MEDIUM] `REDIS_SESSION_DB` placed away from other `REDIS_*` variables**

The other Redis vars are grouped at lines 58-64. This one is added after `REDIS_TLS_KEY`, separated from the core connection vars.

**Suggested fix:** Move it to sit with the other `REDIS_*` connection vars (after `REDIS_AUTH` on line 60, before `REDIS_TLS_ENABLED`).

File: tests/test_api_tool_session_store.py [hunk 8]
```diff
@@ -0,0 +1,328 @@
+"""Unit tests for APIToolSessionStore and session models."""
```
Comment (Thirunayan22) [inline tests/test_api_tool_session_store.py:1]: **[MEDIUM] All tests mock Redis — no integration coverage**

Every test patches `get_redis_client` with a mock. This verifies the mock wiring, not that the actual Redis commands (`SET` with `EX`, `GET`, `DELETE`) behave correctly. Edge cases like TTL expiry, serialization quirks, or connection pool behavior are untested.

**Suggested fix:** Add a small integration test suite using `testcontainers` or a Redis fixture that exercises the real client against an ephemeral Redis instance. Even 3-4 round-trip tests would catch serialization and TTL issues.

File: docs/REDIS_SESSION_STORE.md [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        request.chatId,
+        collected_params=merged_params,
+        turn_count=session.turn_count + 1,
+    )
+    return ask_for_next_missing_param(session.selected_endpoint, merged_params)
+
+# --- Guard: abandon if turn limit reached ---
+if session.turn_count >= session.max_turns:
+    await session_store.delete(request.chatId)
+    return "I was unable to collect all required information. Please try again."
```
Comment (Thirunayan22) [inline docs/REDIS_SESSION_STORE.md]: **[MEDIUM] Dead code in usage example — guard runs after `return`**

This `# --- Guard: abandon if turn limit reached ---` block is unreachable. Both the `if` and `else` branches above it (lines 147-159) end with `return`, so this code never executes. This teaches consumers a wrong pattern.

**Suggested fix:** Move the guard check *before* the `if all_params_collected(...)` block, or restructure to check `turn_count` at the top of the "Turn 2+" section.

File: src/llm_orchestration_service_api.py [hunk 10]
```diff
@@ -92,6 +107,8 @@ async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
         await app.state.orchestration_service.aclose()
         app.state.orchestration_service = None
 
+    await close_redis_client()
```
Comment (Thirunayan22) [inline src/llm_orchestration_service_api.py]: **[LOW] `close_redis_client()` outside try/except**

If this raises, it could prevent the shutdown log or other cleanup from running. The `orchestration_service.aclose()` above is protected by a conditional check, but this Redis close call is bare.

**Suggested fix:**
```python
try:
    await close_redis_client()
except Exception as e:
    logger.warning(f"Error closing Redis client: {e}")
```

File: src/models/session_models.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
+"""Pydantic models for API tool session state."""
+
+from typing import Any, Dict, Optional
```
Comment (Thirunayan22) [inline src/models/session_models.py]: **[LOW] Legacy `typing.Dict` / `typing.Optional` imports**

This project targets Python 3.12 (per `pyproject.toml`). Use `dict[str, Any]` and `X | None` directly instead of importing `Dict` and `Optional` from `typing`.

**Review by Thirunayan22 — CHANGES_REQUESTED** [review]
### Review Overview

**Verdict: Request Changes** — 2 critical issues must be resolved before merge.

#### Issues by Severity

**Critical**
- Hardcoded Redis password in docker-compose files (violates ADR SEC-001)
- Race condition in `update()` — non-atomic read-modify-write

**High**
- No TLS support despite existing TLS config vars in `env.example`
- Silent `None` fallback with no consumer feedback mechanism
- Global mutable state for Redis client singleton

**Medium**
- `update()` accepts arbitrary `**kwargs` with no field validation
- `REDIS_SESSION_DB` misplaced in `env.example`
- No integration tests — all Redis interactions mocked
- Dead code in documentation usage example

**Low**
- `close_redis_client()` outside try/except in shutdown path
- Legacy `typing.Dict` / `typing.Optional` imports (Python 3.12 project)

#### ADR Compliance

| ADR | Status | Notes |
|-----|--------|-------|
| **SEC-001: Secrets Management** | **VIOLATION** | Default password `myredissecret` hardcoded in compose files |
| **ARCH-001: Modular Architecture** | PASS | Clean module separation |
| **VIS-003: Observability Contracts** | PARTIAL | Health check present, no `/metrics` for session ops |
| **DATA-001: Client Data Boundary** | NEEDS REVIEW | Session data may contain user-provided params |
| **NOOPS-004: Fault Tolerance** | PARTIAL | Graceful degradation present, no retry/backoff |
| **LOG-001: Structured Logging** | PARTIAL | Loguru used, not JSON schema per ADR |

File: src/utils/redis_client.py (hunk 5 above)
Comment (nuwangeek) [reply src/utils/redis_client.py:10]: No real concurrency risk in asyncio — this is a single-threaded event loop, so [_redis_client](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html) reference reads/writes are atomic. A coroutine will always see either the client or None, never a torn reference.
ASGI lifespan prevents the shutdown race — [close_redis_client()](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html) runs only after the yield in the lifespan context manager, meaning uvicorn has already stopped accepting new requests and drained in-flight ones before the client is set to None.
Each store method already captures the reference locally ([client = get_redis_client()](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html) at the top), so even if the global changed mid-method, the local binding is safe.
Moving to [app.state](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html) + DI would be a high blast radius refactor — [APIToolSessionStore](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html), [check_redis_health()](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html), and all tests would need to thread [Request](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html) objects through, with no concrete bug being fixed. The module-level connection pool singleton is also the standard pattern recommended by [redis-py](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html)'s own async docs.

File: src/llm_orchestration_service_api.py (hunk 4 above)
Comment (nuwangeek) [reply src/llm_orchestration_service_api.py:97]: Already addressed — [require_session_store](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (FastAPI dependency) was added to [api_tool_session_store.py](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html). Any endpoint that needs multi-turn session state declares [Depends(require_session_store)](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html), which returns HTTP 503 with a clear message when Redis is unavailable. Endpoints that don't need sessions are unaffected. The None fallback at startup is kept intentionally so the wider service (RAG, embeddings) remains available even when Redis is down.

File: src/utils/api_tool_session_store.py (hunk 6 above)
Comment (nuwangeek) [reply src/utils/api_tool_session_store.py:76]: Already addressed — [_VALID_SESSION_FIELDS](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (a [frozenset](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html) derived from [APIToolSession.model_fields](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html)) was added at module level, and [update()](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html) now validates field names before touching Redis. A typo like [sate="ready"](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html) raises [ValueError("Unknown session fields: {'sate'}")](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html) immediately. A test covering this case ([test_update_rejects_unknown_fields](vscode-file://vscode-app/c:/Users/charith.bimsara_root/AppData/Local/Programs/Microsoft%20VS%20Code/41dd792b5e/resources/app/out/vs/code/electron-browser/workbench/workbench.html)) was also added.

**Review by Thirunayan22 — APPROVED** [review]
